Part 7 метапрограммирование

kpkodil opened the pull request.

kpkodil committed:
Скопированы файлы из задания Part_6

kpkodil committed:
Созданы модули.

Созданы модули accessors и validations.
В модуле accessors созданы методы  класса:
 attr_accessor_with_history  - динамическое создание сеттеров и храниние значений в истории.
strong_attr_accessor - проверяет присваемое значение переменной на указанный тип.

В модуле validations:
Метод класса validate - проверка атрибутов класса, согласно заданию.
Инстанс-метод validate! - вызывает все проверки из validate к атрибуту инстанса класса.
Метод реализован через хеш и блок.
- В Хеше повторяются значения атрибута (Не знаю, как реализовать без повторений, чтобы один раз передать это значение в блок each, а из него в метод validate)
- Блок (Пустой. Можно ли так делать?) передается в метод класса validate.

russo-matrosso commented on Part_7/modules/validations.rb:
```diff
@@ -0,0 +1,74 @@
+  module ClassMethods
+    def validate(atr, valid_type, *parameters)
+      if block_given?
+        yield
```
Блок тут точно не нужен

russo-matrosso commented on Part_7/modules/validations.rb:
```diff
@@ -0,0 +1,74 @@
+      else
+        value = class_variable_get("@@#{atr}".to_sym)
+      end
+      case valid_type
```
Просто добавляй в в массив хэш, в котором будет имя валидатора, имя переменной и опции. Потом вызывай на основе этого нужные методы в validate!

russo-matrosso commented on Part_7/modules/validations.rb:
```diff
@@ -0,0 +1,74 @@
+      end
+    end
+
+    def validate_presence(value)
```
Эти методы должны быть методами экземпляра

kpkodil committed:
Изменены методы.

теперь метод класса validate хранит в  себе все варианты проверок, введенных пользователем.

инстанс метод validate! проходит по всем проверкам (тоже инстанс методам, соответствующим типу проверки).

kpkodil committed:
Модули додавлены к классам.

Обработка исключений проводится в интерфейсе.
атрибут train_station в классе Train перенесен в attr_accessor_with_history

russo-matrosso commented on Part_7/modules/validations.rb:
```diff
@@ -7,72 +7,56 @@ def self.included(base)
+        valid_type = validation[:valid_type]
+        parameter = validation[:option]
+        case valid_type
+        when :presence
```
Нужно делать send "validate_#{valid_type}" - иначе будет невозможно расширить список валидаций

kpkodil committed:
Изменен метод validate!

Теперь вызов методов-проверок осуществляется через send.
